File: src/poe2_companion/requirements.py
"""Native requirement provenance; equipment/gem maxima and support totals differ."""
from __future__ import annotations
from typing import Annotated, Literal, TYPE_CHECKING
from pydantic import Field
from .builds import DTO, bounded_dto, BuildError
from .subjects import SubjectBinding

if TYPE_CHECKING:
    from .engine_models import EngineSnapshot

Value = Annotated[float, Field(ge=0, le=1e12, allow_inf_nan=False)]


class Attributes(DTO):
    strength: Value = 0
    dexterity: Value = 0
    intelligence: Value = 0


class RequirementSource(DTO):
    kind: Literal['equipment', 'native_gem', 'support_total']
    saved_item_id: Annotated[int, Field(ge=1, le=1000000)] | None = None
    slot: Annotated[str, Field(max_length=80)] | None = None
    skill_instance_id: Annotated[str, Field(max_length=100)] | None = None
    skill_id: Annotated[str, Field(max_length=120)] | None = None
    support: bool = False
    native_level: Annotated[int, Field(ge=1, le=100)] | None = None
    effective_level: Annotated[int, Field(ge=1, le=1000)] | None = None
    character_level_required: Annotated[int, Field(ge=0, le=1000)] = 0
    base_attributes: Attributes
    modified_attributes: Attributes
    satisfying_attributes: Attributes
    attribute_requirements_ignored: bool
    satisfied: bool


class RequirementBreakdown(DTO):
    available: Attributes
    engine_required: Attributes
    equipment_maximum: Attributes
    native_gem_maximum: Attributes
    support_sum: Attributes
    sources: Annotated[list[RequirementSource], Field(max_length=10000)]
    aggregation: Literal['max_equipment_max_native_gems_sum_supports'] = 'max_equipment_max_native_gems_sum_supports'
    weapon_scope: Literal['pinned_engine_active_skill_set_support_count'] = 'pinned_engine_active_skill_set_support_count'
    own_attributes_prove_initial_equip: Literal[False] = False


class RequirementsRequest(DTO):
    calculation_id: Annotated[str, Field(pattern=r'^calc_[0-9a-f]{32}$')]
    side: Literal['baseline', 'result'] = 'baseline'
    offset: Annotated[int, Field(ge=0, le=10000)] = 0
    limit: Annotated[int, Field(ge=1, le=8)] = 5


class RequirementsPage(DTO):
    calculation_id: str
    side: Literal['baseline', 'result']
    subject: SubjectBinding | None
    summary: RequirementBreakdown
    total: int
    next_offset: int | None
    transition_order: Annotated[list[str], Field(max_length=32)]
    final_equipment_validity: Literal['pass', 'fail', 'indeterminate']
    detail_scope: Literal['native_calculation_requirements_not_live_equipment'] = 'native_calculation_requirements_not_live_equipment'


def page(request: RequirementsRequest, snapshot: EngineSnapshot) -> RequirementsPage:
    from .engine_models import EngineError
    if snapshot.requirements is None:
        raise EngineError('calculation_expired_or_unavailable')
    summary = snapshot.requirements.model_copy(deep=True)
    total = len(summary.sources)
    summary.sources = summary.sources[request.offset:request.offset + request.limit]
    result = RequirementsPage(calculation_id=request.calculation_id, side=request.side,
        subject=snapshot.subject, summary=summary, total=total,
        next_offset=request.offset + len(summary.sources) if request.offset + len(summary.sources) < total else None,
        transition_order=list(snapshot.equip_order),
        final_equipment_validity=snapshot.equipment_validity or snapshot.validation)
    while True:
        try:
            return bounded_dto(result)
        except BuildError:
            if len(summary.sources) <= 1:
                raise
            summary.sources.pop()
            result.next_offset = request.offset + len(summary.sources)

File: src/poe2_companion/lua/requirements.lua
-- Project native requirement tables using canonical identities only.
local M={}
local attrs={Str='strength',Dex='dexterity',Int='intelligence'}
local function zeros() return {strength=0,dexterity=0,intelligence=0} end
function M.project(build)
 local env,out=build.calcsTab.mainEnv,build.calcsTab.mainOutput
 local result={available=zeros(),engine_required=zeros(),equipment_maximum=zeros(),
  native_gem_maximum=zeros(),support_sum=zeros(),sources=setmetatable({},{__jsontype='array'})}
 local ids={}
 for sid,set in pairs(build.skillsTab.skillSets) do
  for gid,group in ipairs(set.socketGroupList or {}) do
   for nid,gem in ipairs(group.gemList or {}) do ids[gem]='skill:s'..sid..':g'..gid..':n'..nid end
  end
 end
 for attr,name in pairs(attrs) do result.available[name]=math.max(0,out[attr] or 0);result.engine_required[name]=out['Req'..attr] or 0 end
 local ignored=not not env.modDB:Flag(nil,'IgnoreAttributeRequirements')
 local highest=env.modDB:Flag(nil,'GemAttributeRequirementsSatisfiedByHighestAttribute')
 local strength=env.modDB:Flag(nil,'StrengthSatisfiesMeleeWeaponsAndSkills')
 for _,req in ipairs(env.requirementsTable or {}) do
  local item,gem=req.sourceItem,req.sourceGem
  local kind=item and 'equipment' or gem and 'native_gem' or 'support_total'
  local row={kind=kind,base_attributes=zeros(),modified_attributes=zeros(),satisfying_attributes=zeros(),
   attribute_requirements_ignored=ignored,satisfied=true}
  local mult=1;local melee=false
  if item then
   row.saved_item_id=not item._companionTrade and item.id or nil;row.slot=req.sourceSlot
   row.character_level_required=tonumber((item.requirements or {}).level) or 0
   mult=item.base and item.base.weapon and out.GlobalWeaponAttributeRequirements or out.GlobalItemAttributeRequirements
   melee=item.base and item.base.weapon and build.data.weaponTypeInfo[item.base.type].melee
   if item.base and item.base.type=='Gloves' and env.modDB:Flag(nil,'IgnoreGloveAttributeRequirements') then mult=0 end
  elseif gem then
   row.skill_instance_id=ids[gem];row.skill_id=gem.skillId;row.native_level=gem.level
   row.support=not not (gem.gemData and gem.gemData.grantedEffect.support)
   row.effective_level=gem.displayEffect and gem.displayEffect.level or nil
   row.character_level_required=gem.reqLevel or 0
   mult=out.GlobalGemAttributeRequirements
   melee=gem.gemData and gem.gemData.tags and gem.gemData.tags.melee
  end
  row.satisfied=row.character_level_required==nil or row.character_level_required<=build.characterLevel
  local bucket=kind=='equipment' and result.equipment_maximum or kind=='native_gem' and result.native_gem_maximum or result.support_sum
  for attr,name in pairs(attrs) do
   local raw=tonumber(req[attr]) or 0
   local modified=math.floor(raw*(mult or 1))
   local available=out[attr] or 0
   if gem and highest then available=math.max(out.Str or 0,out.Dex or 0,out.Int or 0) end
   if attr~='Str' and strength and melee then available=math.max(available,out.Str or 0) end
   row.base_attributes[name]=raw;row.modified_attributes[name]=modified
   row.satisfying_attributes[name]=math.max(0,available)
   if not ignored and modified>available then row.satisfied=false end
   bucket[name]=math.max(bucket[name],modified)
  end
  result.sources[#result.sources+1]=row
 end
 return result
end
return M
